File: 22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python
#
# [22] 括号生成
#

# @lc code=start
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        def foo(l, r, n, cur, res):
            if l == n and r == n:
                res.append(cur)
                return
            if l < n:
                foo(l + 1, r, n, cur + "(", res)
            if r < l:
                foo(l, r + 1, n, cur + ")", res)

        res = []
        foo(0, 0, n, "", res)
        return res
# @lc code=end

File: 22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
public:
    void foo(string current, int left, int right, vector<string>& res) {
        if (left == 0 && right == 0) {
            res.push_back(current);
        }
        if (left > 0) {
            foo(current + "(", left - 1, right, res);
        }
        if (right > left) {
            foo(current + ")", left, right - 1, res);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        foo("", n, n, res);
        return res;
    }
};
// @lc code=end

